### DISCUSSION TITLE
What about a optionally writable atoms?   A `useAtomWithOptionallyWrite` hook?

### DISCUSSION THREAD

**Discussion by davazp** [pmndrs/jotai#1930] [Ideas] [answered]
Right now we have `useAtom` for writeable atoms,  but it returns `never` as the setter for read only atoms.

I would like to be able to distinguish between Atom and WritableAtom at runtime,  by making the setter `undefined` when the atom is not writable.

My use case is that I want to be able to pass both a writable data but also readonly data and allow the component to disable the editing capabilities in a type-safe way.

This is helpful because of the limitation of React of not using conditional hooks

For example, I could define this like

```typescript
export function useAtomWithOptionalWrite<T>(atom: Atom<T> | PrimitiveAtom<T>) {
  const value = useAtomValue(atom);
  const setter = useSetAtom(atom); // ?? undefined   !???
  return [value, setter] as const;
} 
```

But I can't use `useSetAtom` conditionally myself.

**Comment by dai-shi** [comment, marked as answer]
You can conditionally create atoms.

Here's JS version.
```js
const isWritableAtom = (anAtom) => 'write' in anAtom; // needs to understand implementation details
const dummyWritableAtom = atom(null, () => {
  // noop or something you want
});

export function useAtomWithOptionalWrite(anAtom) {
  const value = useAtomValue(anAtom);
  const setter = useSetAtom(isWritableAtom(anAtom) ? anAtom : dummyWritableAtom);
  return [value, setter];
}
```

Ah, you want undefined? You can do a similar approach.

**Reply by davazp** [reply]
Nice tricks!

But can I return undefined from `useSetAtom`? it seems to break the type declaration `SetAtom<Args, Result>`.

I can of course define my own.. maybe like

```typescript
export function useOptionalSetAtom<Value, Args extends unknown[], Result>(
  atom: WritableAtom<Value, Args, Result>
) {
  const store = getDefaultStore();
  const setAtom = useMemo(
    () =>
      "write" in atom
        ? (...args: Args) => {
            return store.set(atom, ...args);
          }
        : undefined,
    [store, atom]
  );
  return setAtom;
}

```

then the trick would work.

**Reply by davazp** [reply]
Oh I see what you meant.  I can have a dummy setter and conditionally return `undefined` or the setter based on the atom 👍  

That will do it!  Thanks!

**Reply by dai-shi** [reply]
That's too tricky. `getDefaultStore` is not ideal there. You should use `useStore`.

Can't we do it without Store API?

```js
export function useAtomWithOptionalWrite(anAtom) {
  const value = useAtomValue(anAtom);
  const setter = useSetAtom(anAtom);
  return [value, isWritableAtom(anAtom) ? setter : undefined];
}
```

**Reply by davazp** [reply]
For reference, this is the working version I got for now

```typescript
export function useAtomWithOptionalWrite<T>(
  atom: Atom<T> | PrimitiveAtom<T>
): [Awaited<T>, SetAtom<[SetStateAction<T>], void> | undefined] {
  const value = useAtomValue(atom);

  const setter = useSetAtom(
    "write" in atom ? atom : (dummySetter as unknown as PrimitiveAtom<T>)
  );

  return [value, "write" in atom ? setter : undefined];
}
```

It is still lying to Typescript on the dummy setter, but it is hopefully not observable from outside.

**Reply by dai-shi** [reply]
In that sense, doesn't this work?
```ts
  const setter = useSetAtom(atom as any);
```

**Reply by davazp** [reply]
Maybe?  You know better. But I think this relies more on internals.

I can imagine useSetAtom throwing if atom is not writable or something like that.

In any case, it works for now. Thanks for the help!

**Reply by dai-shi** [reply]
Yeah, that's true. Either should be good if it works and both rely on internals to some extent.
